File: src/main/java/com/baulsupp/oksocial/tracing/ZipkinConfig.java
package com.baulsupp.oksocial.tracing;

import brave.propagation.TraceContext;
import java.io.File;
import java.io.FileReader;
import java.io.IOException;
import java.util.Optional;
import java.util.Properties;
import java.util.function.Function;
import javax.annotation.Nullable;

import static java.util.Optional.ofNullable;

public class ZipkinConfig {
  private @Nullable String senderUri;
  private @Nullable String displayUrl;
  private static File zipkinRc = new File(System.getenv("HOME"), ".zipkinrc");

  public ZipkinConfig(String senderUri, String displayUrl) {
    this.senderUri = senderUri;
    this.displayUrl = displayUrl;
  }

  public static ZipkinConfig load() throws IOException {
    if (zipkinRc.exists()) {
      try (FileReader r = new FileReader(zipkinRc)) {
        Properties p = new Properties();
        p.load(r);
        String sender = p.getProperty("SENDER");
        String display = p.getProperty("DISPLAY");
        return new ZipkinConfig(sender, display);
      }
    } else {
      return unconfigured();
    }
  }

  public static ZipkinConfig unconfigured() {
    return new ZipkinConfig(null, null);
  }

  public Optional<String> zipkinSenderUri() {
    return ofNullable(senderUri);
  }

  public Function<TraceContext, Optional<String>> openFunction() {
    return traceContext -> ofNullable(displayUrl).map(
        url -> url.replaceAll("\\{traceid\\}", traceContext.traceIdString()));
  }
}

File: src/main/kotlin/com/baulsupp/oksocial/tracing/ZipkinConfig.kt
package com.baulsupp.oksocial.tracing

import brave.propagation.TraceContext
import java.io.File
import java.io.FileReader
import java.io.IOException
import java.util.Properties

class ZipkinConfig(private val senderUri: String?, private val displayUrl: String?) {

    fun zipkinSenderUri(): String? {
        return senderUri
    }

    fun openFunction(): (TraceContext) -> String? {
        return { traceContext -> displayUrl?.replace(Regex("""\{traceid\}"""), traceContext.traceIdString()) }
    }

    companion object {
        private val zipkinRc = File(System.getenv("HOME"), ".zipkinrc")

        @Throws(IOException::class)
        fun load(): ZipkinConfig {
            if (zipkinRc.exists()) {
                FileReader(zipkinRc).use { r ->
                    val p = Properties()
                    p.load(r)
                    val sender = p.getProperty("SENDER")
                    val display = p.getProperty("DISPLAY")
                    return ZipkinConfig(sender, display)
                }
            } else {
                return unconfigured()
            }
        }

        fun unconfigured(): ZipkinConfig {
            return ZipkinConfig(null, null)
        }
    }
}
